### ISSUE TITLE
RFC: Avatax & discounts compatibility

### ISSUE THREAD

**Issue by zedzior** [saleor/saleor#16595]
## Problem

The Avatax app does not have enough information on how to distribute order-level discounts over the lines, resulting in incorrect tax calculations. 

Saleor sends taxation data to Avatax in a form of `TaxableObject` type, which contains all order level discounts inside a single list with two attributes only:`name` and ` amount`. Therefore Avatax doesn't have information if the discount should reduce the value of subtotal, shipping, or both. 

Currently, it distributes all the discounts in the same way: proportionally reduce all the lines including shipping (in Avalara, shipping is treated as an order line). 
Saleor offers the following order-level discounts:
- manual (applied to both subtotal and shipping price)
- entire order voucher (applied to subtotal)
- shipping voucher (applied to shipping price)
- order promotion on subtotal (applied to subtotal)

## 1. General Assumptions

1. Do not allow discounting both subtotal and shipping with a single discount. This should be always handled separately. It currently concerns manual order level discounts only.
2. Avatax should apply all `TaxableObject.discounts` to subtotal only, meaning omitting the line responsible for shipping
3. To apply a shipping discount we will add a separate field to `TaxableObject` containing shipping discounts only. Those will be applied to the line responsible for shipping. Currently, it will be only a shipping portion of the manual discount.  
4. Do not change the logic for line-level discounts (including shipping vouchers). It should be as it is, already applied to the lines.
5. To keep backward compatibility, under the hood, we should split the manual discount into `manual subtotal discount` and `manual shipping discount` using proportions based on values before taxation. Going forward we should deprecate the `orderDiscountAdd` mutation and substitute it with two separate mutations: `orderDiscountSubtotalAdd` and `orderDiscountShippingAdd`
6. Shipping discounts, same as subtotal discounts, should be associated with order/checkout, distinguishable by new `DiscountTypes`: `MANUAL_SUBTOTAL` and `MANUAL_SHIPPING`
7. Existing manual discounts should be split with migration
8. The Avatax app will be adjusted as follows (no calculations on the app side):
  - If `TaxableObject.discounts` contains any objects, mark all lines except shipping as `discounted=True` 
  - If `TaxableObject.shipping_discounts` contains any objects, add as many lines with negative values and mark the lines with the same `taxCode` as shipping line (please refer to: https://developer.avalara.com/erp-integration-guide/sales-tax-badge/transactions/discounts-and-overrides/discounting-a-transaction/#:~:text=Negative%20amount%20line%20items)
9. Update plugin logic to handle both shipping and subtotal manual discounts 

### Types

```graphql
class TaxableObject(BaseObjectType):
    …
    discounts = NonNullList(   # existing field for reference
        TaxableObjectDiscount, 
        description="List of discounts applicable to subtotal.",  # new description
        required=True
    )
    shipping_discounts = NonNullList(   # new field
        TaxableObjectDiscount, 
        description="List of discounts applicable to shipping.", 
        required=True
    )
    …

enum OrderDiscountType {
    …
    MANUAL_SUBTOTAL
    MANUAL_SHIPPING

```

## 3. Database changes
No changes to database

## 4. UML diagrams
No need 

## 5. To Do list
```[tasklist]
- [ ] Add `shipping_discounts` field to `TaxableObject`
- [ ] Rewrite `orderDiscountAdd`mutation to create two discounts for shipping and subtotal
- [ ] Adjust logic responsible for applying and updating manual discounts 
- [ ] Migrate existing order level manual discounts into shipping and subtotal discounts
```

## 6. To Test
TBD

## 7. To investigate
1. What happens with the remaining discount portion when discount is not applicable to a specific line?
2. Is it legal from a tax law perspective, that we apply line level discounts to all the lines by ourselves, no matter what the line's `taxCode` is?

## 8. References

1. https://quaderno.io/blog/sales-tax-discounted-goods/
2. https://developer.avalara.com/erp-integration-guide/sales-tax-badge/transactions/discounts-and-overrides/discounting-a-transaction/
3. https://www.figma.com/board/Zx1I6WGblKIEGMrr5VUstU/Discounts-with-AvaTax-app-refi?node-id=0-1&t=zLKIwVsM5gZahJe0-0

**Comment by korycins** [comment]
> 7. Existing manual discounts should be split with migration

We could think how to handle this. I mean:
- Do we want to always split the discount? - For example if we have MANUAL-DISCOUNT for 10USD which is lower than subtotal(99), then we could just rename it to MANUALSUBTOTAL. I am pointing this in terms of data that we need to process. Maybe this way would be faster, but we can come back to this during implementation I guess.
- Do we want to do that for all orders (including the one which are already confirmed and nothing can change there)?

---

Would be also good to have a plan for 0-downtime migration. I mean if you want to fix the problem in 3.20 and someone is already on 3.20, then the conversion of MANUAL_DISCOUNT into MANUAL_SUBTOTAL/MANUAL_SHIPPING can be problematic - as both of the discounts will be present. I am just raising this, to think about this before starting the work. 

---

> 1.What happens with the remaining discount portion when discount is not applicable to a specific line?

This would be handled by Saleor, right? I mean new mutation should not allow you to provide a discount greater than SUBTOTAL/SHIPPING.  But if you think about this from app perspective, then I don't think this is a scope of this RFC. I mean, we should just confirm that we don't allow you to provide a greater discount than the current amount

> 2. Is it legal from a tax law perspective, that we apply line level discounts to all the lines by ourselves, no matter what the line's taxCode is?

The proposed changes are related to fixing the existing problem that we have which is: manual propagated on subtotal and shipping.  This question could be addressed in separate RFC, where we will introduce the API that would provide only base price + discounts, right?

**Comment by zedzior** [comment]
It was not implemented in favour of: https://github.com/saleor/saleor/issues/16613
